### PR TITLE
CPR-509 add support for metadata files

### PR DESCRIPTION
**Pull request by e-gris**
- Add support for traversing trees of file metadata
- tweaked logging so it prints the script information
- try to make variables and functions descriptive
- update to version 0.2.0

### CODE REVIEW

File: README.rst
Comment (mwaggett) [inline README.rst]: `--help` for consistency with the rest of the optional args?

File: README.rst
Comment (mwaggett) [inline README.rst]: perhaps reverse the order of these, for consistency

File: README.rst
Comment (mwaggett) [inline README.rst]: What's the default?

File: README.rst
Comment (mwaggett) [inline README.rst]: you lost the `[-h]` altogether

Comment (e-gris) [reply README.rst]: Oopsies.

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py:64]: I think you're missing a `fi` and an `end` to the `def`.

Comment (e-gris) [reply generate_directory_indexes.py:64]: Nope. This is Python. No 'end' statements.

Comment (mwaggett) [reply generate_directory_indexes.py:64]: oh right. ugh. It's been a while since I python'd.

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py:86]: I know this wasn't part of your changeset, but couldn't this just be `path = prefix.replace(base_path, '')`?

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py]: What is `k`?

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py]: extra space in there, it looks like

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py]: Does python support question marks in function defs?

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py]: couldn't you just `return file_name in index_file_names`?

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: i, j, & k are conventionally reserved generic index variables for tight loops. I left these here.

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: It doesn't.

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: Wrong direction. The intent here is to provide a condition for not including certain files in the index.html files, namely the index.html files themselves. Previously there was a regex for that that I replaced with this whitelist. 

IOW, does current file match a special file? If Yes, then skip it.

A refactoring in the future will combine the logic of generating the index file name with creating the skippable-list. However, after some thought, that felt outside the scope of this already change-heavy PR.

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py:86]: Quite possibly. I'd have to tinker with it to see if that excludes some edge case that isn't obvious. Left alone for now.

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py]: Is the reason for this ordering having non-default behavior first? because to me, having `if ... != ...` feels weird, but I do understand your reasoning.

Comment (e-gris) [reply generate_directory_indexes.py]: Just a general pattern that I use of handling exceptional/edge/non-default cases first. This calls attention to the fact that they are exceptional and helps with preventing bugs.

File: generate_directory_indexes.py
Comment (mwaggett) [reply generate_directory_indexes.py]: I understand how you're using this functions, but I'm pretty sure, if you zoom out, this just says
```
if (true)
   return True
return False
```
unless I'm just wrong about what python's `<blah> in <blah_blah>` does..

Comment (e-gris) [reply generate_directory_indexes.py]: file_name will rarely be in index_file_names. It will likely be something like foo-bar-1.0.rpm. This function will return False most of the time.

Comment (mwaggett) [reply generate_directory_indexes.py]: yeah, but my point is that this could just be simplified.
so sometimes, `return (true)`, but usually `return (false)`.
Am I making sense?

Comment (e-gris) [reply generate_directory_indexes.py]: I could use:

```python
return True if file_name in index_file_names else False
```
but I don't find that terribly readable.

Comment (e-gris) [reply generate_directory_indexes.py]: Now I understand your original comment. I misread it originally. You're right, I could use:

```python
return file_name in index_file_names
```
It *is* more succinct but isn't as obvious IMO.

Comment (mwaggett) [reply generate_directory_indexes.py]: yeah, that's what I meant. There's definitely a balance between readability and succinctness, so I'll let you make the call.

**Review by mwaggett — APPROVED** [review]
assuming you've tested this somehow? 👍

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: After thinking about it overnight your version, I think, is correct. There's another tweak that I want to try as well.

**Comment by mwaggett** [comment]
should probably squash the commits before merge, but otherwise lgtm 😊

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: We should use `"""` for comment blocks instead of `## `

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: Same `"""` comment block issue as above. We should also move this comment into the function.
```py
def render_index(prefix, order_by, contents, reverse_order, base_path):
    """
    Use a templating library to turn a prefix and a list of contents
    into an HTML directory index
    """
```

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: Same `"""` comment issue as above.

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: Don't insert spaces between the `=` in argument lists when setting argument defaults.

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: Nitpicky but you should use `is not` instead of `!=` when testing equality with `None`.

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: We should push these methods into some sort of boilerplate for python so that you can easily debug the file in a repl.

Try this instead:
```py
if __name__ == '__main__':
    parse_arguments()
    validate_input(configuration)
    traverse_tree(configuration.path, configuration.path, configuration.file_metadata)
```

File: generate_directory_indexes.py
Comment (demophoon) [reply generate_directory_indexes.py]: On second thought this could probably be simplified to just `if file_metadata`.

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: Since we are not mutating `script` in this function this is unnecessary.

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: If possible we should be using `.format()` instead of `%` for formatting strings as `%` is deprecated.

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: Is there any reason why `script` is in the top-level scope? I'm only seeing it be used in `validate_input`. We could push this down into that function's scope to declutter top-level scope as much as possible.

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: I like the explicit "is not"

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: There *was* a reason for it but it's no longer applicable.

File: generate_directory_indexes.py
Comment (mwaggett) [inline generate_directory_indexes.py:23]: probably want to remove this too?

File: generate_directory_indexes.py
Comment (demophoon) [inline generate_directory_indexes.py]: ~Nitpick but we should use `is` and not `==` when testing equality with `None`.~

Sorry. not this.. explanation comment coming soon.

File: generate_directory_indexes.py
Comment (demophoon) [reply generate_directory_indexes.py]: Other than reading better it tests object identity instead of object equality. For things like `None`, `False`, and `True` we don't want to worry about what `__eq__` (or `==`) is going to do to test equality between the objects. With `is` and `is not` we actually test if the literal `None` we should expect is the exact same as the builtin `None` we are expecting.

However I believe that what we actually want is `if file_metadata:` instead. If `file_metadata` is ever `False` then this will equate to true resulting in the wrong thing happening.

File: generate_directory_indexes.py
Comment (demophoon) [reply generate_directory_indexes.py]: See comment above for explanation of testing equality to `None`.

File: generate_directory_indexes.py
Comment (e-gris) [reply generate_directory_indexes.py]: Got it. Thanks.

**Review by demophoon — APPROVED** [review]
This looks good to me

**Comment by mwaggett** [comment]
@e-gris I will merge once you squash, unless you also want @danieldreier to take a look.

**Comment by e-gris** [comment]
Let's not do this on a Friday afternoon. I'll squash again on Monday.

**Pull request merged**
